src/core/registration.c

```diff
@@ -247,6 +249,18 @@ MsQuicRegistrationShutdown(
             Entry = Entry->Flink;
         }
 
+        CxPlatDispatchLockAcquire(&Registration->ListenerLock);
+        Entry = Registration->Listeners.Flink;
+        while (Entry != &Registration->Listeners) {
+
+            QUIC_LISTENER* Listener =
+                CXPLAT_CONTAINING_RECORD(Entry, QUIC_LISTENER, RegistrationLink);
+
+            MsQuicListenerClose((void *)Listener);
```

Comment:
You should call `MsQuicListenerStop` instead. The app should **always** be the one to explicitly call Close.

Reply:
You are right. But for already stopped listener, calling ListenerStop will not trigger the stop complete callback as it already stopped and that registaion close will still hang because their is no way to trigger the Listener close.

Reply:
That is the same thing for Connections. If the app has gotten a completion event for an object, but not closed/freed it, then it is an app issue; not our issue.